# probevelo

Work-around for RNA velocity inference from 10X Fixed RNA Profiling using probe sets.

## Installation

Only available in experimental stage on GitHub for now:

```bash
pip install git+https://github.com/mvfki/probevelo.git
```

## Usage

### CLI command

After installing the package, you can use the command line interface to count spliced and unspliced reads from a BAM file generated by 10X Fixed RNA Profiling.

```bash
probevelo \
    /path/to/10x/out/sample_alignment.bam \
    /path/to/10x/out/probe_set.csv \
    --output /path/to/output/splicing.h5ad
```

Here's the all the options available:

```
usage: probevelo [-h] [-v] [-t THREADS] [--output OUTPUT] BAM PROBE_SET

ProbeVelo: Work around for counting spliced and unspliced reads from FRP BAM files.

positional arguments:
  BAM                   Path to the BAM file containing RNA sequencing data. Must be indexed.
  PROBE_SET             Path to the probe set CSV file.

options:
  -h, --help            show this help message and exit
  -v, --version         Show program's version number and exit
  -t THREADS, --threads THREADS
                        Number of threads to use for counting (default: 1).
  --output OUTPUT       Output file path for the results (default: output.h5ad).
```

Note that **the BAM file must be indexed**, e.g. you must have `sample_alignment.bam.bai` in the same directory as the input file `sample_alignment.bam`.

### Python API

If you prefer to use the package programmatically, you can do so by importing the `probevelo` module in your Python script. The package provides a simple interface to load the probe set information and BAM information and perform the counting, returning an `anndata.AnnData` object with the results.

```python
import probevelo

probe_set = probevelo.probe_set('/path/to/10x/out/probe_set.csv')
bam_parser = probevelo.bam_parser('/path/to/10x/out/sample_alignment.bam')

adata = bam_parser.count_splice(probe_set, adata=True)
```

The returned `anndata.AnnData` object will contain:

- `adata.X`: the total raw counts, i.e. sum of spliced and unspliced counts
- `adata.layers['spliced']`: the spliced counts
- `adata.layers['unspliced']`: the unspliced counts
- `adata.obs`: Only the barcodes as the index
- `adata.var`: All gene IDs available in the probe set will be used as the index,
and the `gene_name` column will be mapped as an alternative column. Note that multiple
gene IDs can map to the same gene name, so we by default use the ID as the index.
